### PR TITLE
Integrate renter/metaserver. Bug fixes.

### PR DESCRIPTION
**Pull request by AlexSteele**
- Fix provider auth endpoints.
- Save renter contract/file metadata with metaserver (but cache locally
as well)
- If a block cannot be deleted, add it to a list of blocks to delete
later.

### CODE REVIEW

File: renter/upload.go [hunk 1]
```diff
@@ -10,28 +10,184 @@ import (
...
+		ID:         fileId,
+		OwnerID:    r.Config.RenterId,
+		Name:       destPath,
+		IsDir:      false,
+		AccessList: []core.Permission{},
+		AesKey:     base64.URLEncoding.EncodeToString(aesKeyEncrypted),
+		AesIV:      base64.URLEncoding.EncodeToString(aesIVEncrypted),
+		Versions:   []core.Version{},
+	}
+	// TODO: Kincaid. Do I have to set this?
```
Comment (Caid11) [inline renter/upload.go]: No. The server will set it. If you set it, it'll just override it.

Comment (Caid11) [reply renter/upload.go]: It does that so if more than one client uploads a version at the same time, they'll get unique version numbers (it uses Mongo's atomic `$inc` operator to get the version number).

File: renter/download.go [hunk 2]
```diff
@@ -23,11 +23,14 @@ import (
 	"github.com/klauspost/reedsolomon"
 )
 
-func (r *Renter) Download(fileId string, destPath string) error {
+func (r *Renter) Download(fileId string, destPath string, versionNum *int) error {
```
Comment (Caid11) [inline renter/download.go]: Is that... a pointer to an int? I'm curious as to why you're using that.

File: renter/download.go [hunk 3]
```diff
@@ -340,3 +349,13 @@ func convertToReaderSlice(files []*os.File) []io.Reader {
 	}
 	return res
 }
+
+
+func findVersion(file *core.File, versionNum int) *core.Version {
+	for i := 0; i < len(file.Versions); i++ {
```
Comment (Caid11) [inline renter/download.go]: You can fetch versions directly from the metaserver using the endpoint `/files/{id}/versions/{id}` or something to that effect (it's in the metaclient). Could be useful if you want the same behavior I used for getting shared files (check first in cache, then check metaserver).

File: renter/download.go (hunk 2 above)
Comment (AlexSteele) [reply renter/download.go]: Weak type system: This is the most straightforward way to make versionNum optional (in both the requestJSON and Download function).

Comment (Caid11) [reply renter/download.go]: Ah, gotcha.

**Pull request merged**